map_test.go

```diff
@@ -1920,27 +2009,54 @@ func BenchmarkPerCPUMarshalling(b *testing.B) {
 		Type:       PerCPUHash,
 		KeySize:    8,
 		ValueSize:  8,
-		MaxEntries: 1,
+		MaxEntries: uint32(batch),
 	})
 	if err != nil {
 		b.Fatal(err)
 	}
 
 	b.Cleanup(func() { m.Close() })
-	if err := m.Put(key, val[0:]); err != nil {
-		b.Fatal(err)
+	for i := 0; i < batch; i++ {
+		key := uint64(i + 1)
+		if err := m.Put(key, val[0:]); err != nil {
+			b.Fatal(err)
+		}
 	}
 
-	b.Run("reflection", func(b *testing.B) {
+	b.Run("Lookup", func(b *testing.B) {
 		b.ReportAllocs()
 		b.ResetTimer()
 
 		var value []uint64
 
 		for i := 0; i < b.N; i++ {
-			err := m.Lookup(unsafe.Pointer(&key), &value)
-			if err != nil {
-				b.Fatal("Can't get key:", err)
+			for j := 0; j < batch; j++ {
+				key := uint64(j + 1)
+				err := m.Lookup(unsafe.Pointer(&key), &value)
+				if err != nil {
+					b.Fatal("Can't get key:", err)
+				}
+			}
+		}
+	})
+	b.Run("BatchLookup", func(b *testing.B) {
+		b.ReportAllocs()
+		b.ResetTimer()
+
+		var (
+			keys   = make([]uint64, batch)
+			values = make([]uint64, batch*possibleCPU)
+		)
+
+		for i := 0; i < b.N; i++ {
+			count, err := m.BatchLookup(&BatchCursor{}, keys, values, &BatchOptions{})
+			if count <= 0 {
+				if err != nil {
+					b.Fatal("Can't get batch:", err)
+				}
+			}
+			if count != batch {
+				b.Fatalf("Can't get batch: %d", count)
 			}
```

Comment:
I think this benchmark is now comparing two separate things: a direct lookup via m.Lookup and iteration via (the now badly named) `BatchLookup`. The correct thing to compare is `m.Iterate()` and `BatchLookup`, see `BenchmarkIterate`. My suggestion would be to copy that instead of modifying `BenchmarkPerCPUMarshalling`. (See also https://github.com/cilium/ebpf/pull/1254/commits/df14fdedddb5a0365781630efc588c0a690d4a0e) Crucial point: you always need to check for ErrKeyNotExist to figure out whether you are done iterating.

Reply:
Because mostly the syscall cost is likely to be the high cost.

Reply:
```
BenchmarkPerCPUMarshalling/Lookup-4         	    1740	    677886 ns/op	   98304 B/op	    4096 allocs/op
BenchmarkPerCPUMarshalling/Iterate-4        	    1083	   1098589 ns/op	   90210 B/op	    3076 allocs/op
BenchmarkPerCPUMarshalling/BatchLookup-4    	    4807	    252948 ns/op	   65592 B/op	    1029 allocs/op```

Reply:
ok I can copy/attempt refactor of `BenchmarkIterate`